File: P2A4_Herencias/C#/PruebaSistemaNomina.cs
using System;

class PruebaSistemaNomina
{
    static void Main(string[] args)
    {
        EmpleadoAsalariado empleadoAsalariado = new EmpleadoAsalariado("John", "Smith", "111-11-1111", 800.00);
        EmpleadoPorHoras empleadoPorHoras = new EmpleadoPorHoras("Karen", "Price", "222-22-2222", 16.75, 40);
        EmpleadoPorComision empleadoPorComision = new EmpleadoPorComision("Sue", "Jones", "333-33-3333", 10000, 0.06);
        EmpleadoBaseMasComision empleadoBaseMasComision = new EmpleadoBaseMasComision("Bob", "Lewis", "444-44-4444", 5000, 0.04, 300);
        TrabajadorPiezas trabajadorPiezas = new TrabajadorPiezas("Jane", "Doe", "555-55-5555", 2.50, 200);

        Console.WriteLine("Empleados procesados por separado:\n");

        Console.WriteLine($"{empleadoAsalariado}\ningresos: {empleadoAsalariado.Ingresos():C}\n");
        Console.WriteLine($"{empleadoPorHoras}\ningresos: {empleadoPorHoras.Ingresos():C}\n");
        Console.WriteLine($"{empleadoPorComision}\ningresos: {empleadoPorComision.Ingresos():C}\n");
        Console.WriteLine($"{empleadoBaseMasComision}\ningresos: {empleadoBaseMasComision.Ingresos():C}\n");
        Console.WriteLine($"{trabajadorPiezas}\ningresos: {trabajadorPiezas.Ingresos():C}\n");

        Empleado[] empleados = new Empleado[5];
        empleados[0] = empleadoAsalariado;
        empleados[1] = empleadoPorHoras;
        empleados[2] = empleadoPorComision;
        empleados[3] = empleadoBaseMasComision;
        empleados[4] = trabajadorPiezas;

        Console.WriteLine("Empleados procesados en forma polimorfica:\n");

        foreach (Empleado empleadoActual in empleados)
        {
            Console.WriteLine(empleadoActual);
            
            if (empleadoActual is EmpleadoBaseMasComision)
            {
                EmpleadoBaseMasComision e = (EmpleadoBaseMasComision)empleadoActual;
                e.SalarioBase *= 1.10;
                Console.WriteLine($"nuevo salario base con 10% de aumento es: {e.SalarioBase:C}");
            }
            
            Console.WriteLine($"ingresos {empleadoActual.Ingresos():C}\n");
        }

        for (int i = 0; i < empleados.Length; i++)
        {
            Console.WriteLine($"El empleado {i} es un {empleados[i].GetType().Name}");
        }
    }
}

File: P2A4_Herencias/Python/PruebaSistemaNomina.py
from EmpleadoAsalariado import EmpleadoAsalariado
from EmpleadoPorHoras import EmpleadoPorHoras
from EmpleadoPorComision import EmpleadoPorComision
from EmpleadoBaseMasComision import EmpleadoBaseMasComision
from TrabajadorPiezas import TrabajadorPiezas

def main():
    empleado_asalariado = EmpleadoAsalariado("John", "Smith", "111-11-1111", 800.00)
    empleado_por_horas = EmpleadoPorHoras("Karen", "Price", "222-22-2222", 16.75, 40)
    empleado_por_comision = EmpleadoPorComision("Sue", "Jones", "333-33-3333", 10000, 0.06)
    empleado_base_mas_comision = EmpleadoBaseMasComision("Bob", "Lewis", "444-44-4444", 5000, 0.04, 300)
    trabajador_piezas = TrabajadorPiezas("Jane", "Doe", "555-55-5555", 2.50, 200)

    print("Empleados procesados por separado:\n")
    print(f"{empleado_asalariado}\ningresos: ${empleado_asalariado.ingresos():,.2f}\n")
    print(f"{empleado_por_horas}\ningresos: ${empleado_por_horas.ingresos():,.2f}\n")
    print(f"{empleado_por_comision}\ningresos: ${empleado_por_comision.ingresos():,.2f}\n")
    print(f"{empleado_base_mas_comision}\ningresos: ${empleado_base_mas_comision.ingresos():,.2f}\n")
    print(f"{trabajador_piezas}\ningresos: ${trabajador_piezas.ingresos():,.2f}\n")

    empleados = [
        empleado_asalariado,
        empleado_por_horas,
        empleado_por_comision,
        empleado_base_mas_comision,
        trabajador_piezas
    ]

    print("Empleados procesados en forma polimorfica:\n")
    for empleado_actual in empleados:
        print(empleado_actual)
        if isinstance(empleado_actual, EmpleadoBaseMasComision):
            empleado_actual.salario_base *= 1.10
            print(f"nuevo salario base con 10% de aumento es: ${empleado_actual.salario_base:,.2f}")
        print(f"ingresos ${empleado_actual.ingresos():,.2f}\n")

    for i, empleado_actual in enumerate(empleados):
        print(f"El empleado {i} es un {empleado_actual.__class__.__name__}")

if __name__ == "__main__":
    main()
